Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java

```diff
@@ -717,42 +719,62 @@ private static double getSurfaceArea(final List<Point3f> points) {
 	 * Get the Feret diameter of a surface. Uses an inefficient brute-force
 	 * algorithm.
 	 *
-	 * @param surfacePoints points of all the particles.
-	 * @return Feret diameters of the surfaces.
+	 * @param surfacePoints points from a surface mesher
+	 * @throws IllegalArgumentException if there are less than 12 points (a tetrahedron)
+	 * @return Feret diameters and x, y, z coordinates of the two feret points of each surface,
+	 * packed in a double so that the feret diameter of particle i is found at [i * 7] and the
+	 * points' coordinates are in the following 6 positions in ax, ay, az, bx, by, bz order.
 	 */
-	static double[] getFerets(final List<List<Point3f>> surfacePoints) {
+	static double[][] getFerets(final List<List<Point3f>> surfacePoints) {
 		Thread[] threads = Multithreader.newThreads();
 		final int nSurfaces = surfacePoints.size();
-		final double[] ferets = new double[nSurfaces];
+		//distance, xa, ya, za, xb, yb, zb
+		final double[][] ferets = new double[nSurfaces][7];
 		AtomicInteger ai = new AtomicInteger(0);
 		for (int thread = 0; thread < threads.length; thread++) {
 			threads[thread] = new Thread(() -> {
 				for (int i = ai.getAndIncrement(); i < nSurfaces; i = ai.getAndIncrement()) {
 					final List<Point3f> surface = surfacePoints.get(i);
-				
+					
+					
 					if (surface == null) {
-						ferets[i] = Double.NaN;
+						Arrays.fill(ferets[i], Double.NaN);
 						continue;
 					}
 					
+					final int nPoints = surface.size();
+					
+					//4 triangles * 3 points for the minimal tetrahedron
+					if (nPoints < 12) {
+						Arrays.fill(ferets[i], Double.NaN);
+						throw new IllegalArgumentException("Mesh for particle "+i+" has too few points: "+nPoints);
+					}
+					
 					// check all the point pairs in this surface
-					ListIterator<Point3f> ita = surface.listIterator();
-					ListIterator<Point3f> itb;
-					Point3f a;
-					Point3f b;
+					Point3f a = new Point3f();
+					Point3f b = new Point3f();
```

Comment:
Redundant initializations, can be local variables within the loops

Reply:
Agree this is neater but I worry about object creation inside hot loops. No time right now to benchmark it. Hoping that the JRE will optimise it...

Reply:
There's no object creation in `Point a = surface.get(p)`. A new reference variable `a` is declared, and then a reference to an object in `surface` is assigned to it. It's `Point a = new Point()` where object creation happens.